Add usbtmc class driver (READY FOR TESTING)

pigrew opened the pull request:
Here is an initial revision of a USBTMC class driver.

It has most of the features I need, but not all of the features that exist in USBTMC. I added some comments on what's lacking in the headers of `src/class/usbtmc/usbtmc_device.c`.  It's missing many of the "required" specifications, but is implemented enough to meet most needs.

The example application does not have any message/passing or multi-thread capability at the moment. I'm not planning to use a usbtmc task in the example, but I will in my more final project.

I don't intend on this being accepted right away, as I'm likely to create some breaking changes in the interface, but I want to get feedback on what I have now.

Note that it's against the "develop" branch because it requires the class-type EP-recipient patch to work.

This has been tested with NI MAX's visa control panel on Windows 10. Send, receive, query, and read STB all work (or at least with very short messages which won't overflow buffers).

(I fear I'm created this PR too soon; can delete if I'm committing to it too much, or can just start pushing once a day or something. I also understand why I see so many incomplete USBTMC implementations in commercial hardware.)

Some other implementations:

[karlp/discotmc ](https://github.com/karlp/discotmc/blob/master/device/stm32_ocm3/)
[FTDI FT51A](https://www.ftdichip.com/Support/Documents/AppNotes/AN_347_FT51A_Test_and_Measurement_Sample.pdf)
[Linux USBTMC host class](https://github.com/torvalds/linux/blob/master/drivers/usb/class/usbtmc.c)

pigrew commented on src/class/usbtmc/usbtmc_device.c:
```diff
@@ -0,0 +1,420 @@
+      case USBTMC_MSGID_VENDOR_SPECIFIC_IN:
+      case USBTMC_MSGID_USB488_TRIGGER:
+      default:
+        TU_VERIFY(false);
```
To NAK, should it TU_VERIFY(false) or return false?

pigrew commented on src/class/usbtmc/usbtmc.h:
```diff
@@ -0,0 +1,267 @@
+	usbtmc_msg_header_t header ; ///< Header
+	uint32_t TransferSize      ; ///< Transfer size; LSB first
+	struct {
+      uint8_t EOM  : 1         ; ///< EOM set on last byte
```
C99 bitfields must be unsigned int, signed int, or int. Nothing else. Using uint8_t is a compiler extension.

ladyada commented:
neat! ive never heard of this

pigrew commented:
> neat! ive never heard of this

@ladyada: Yeah, it's a quite powerful protocol. Most "test instruments" like DMM or oscilloscopes speak it over USB. It's designed to replace GPIB (and have all of its features, where ever possible).  I've been wanting to use it it in personal projects for a while, but have not found any complete open-source implementations. My feeling is that each company (Keysight, Tek, LeCroy, etc likely have their own implementations, and these days all of those would be on an embedded Linux or Windows SoC).

This USBTMC spec is much more complicated to implement than I had anticipated. My code in this request (at the moment) is a bit too simplistic (and the number of lines will probably double or triple :/). I've finished up implemented the "indicator pulse" LED flash, but the transaction aborts and device clears are needing a lot of planning (due to the asynchronous nature of them). 

I'm still aiming to have minimal locking, and such. I don't want it to implement its own task, so some locking will be necessary as it can be called both from the USBD core and application code.

I also ran into a few implementation problems with my stm32 fsdev driver which I'm also fixing (it didn't properly support endpoints != 64 bytes, and had a typo-bugs, and had 400 B of type conversions going into its flash). Those should be put together into a PR early this week.

I'll update this PR in a day or once I figure out how to cope with concurrency and aborting transactions. Some progress (both USBTMC and stm fsdev driver) are [my fork](https://github.com/pigrew/tinyusb/tree/stm32f0).

hathach commented on examples/device/usbtmc/src/usb_descriptors.c:
```diff
@@ -0,0 +1,254 @@
+char const* string_desc_arr [] =
+{
+  (const char[]) { 0x09, 0x04 }, // 0: is supported language is English (0x0409)
+  "TinyUSB",                     // 1: Manufacturer
```
yeah, you should remove all other non usbtmc

hathach commented on examples/device/usbtmc/src/tusb_config.h:
```diff
@@ -0,0 +1,66 @@
+
+#define CFG_TUD_USBTMC           1
+#define CFG_TUD_USBTMC_ENABLE_INT_EP
+//#define USBTMC_CFG_ENABLE_488    0
```
if `USBTMC_CFG_ENABLE_488` is user configurable, it should be prefixed with CFG_TUD_USBTMC_

hathach commented on examples/device/usbtmc/src/usbtmc_app.c:
```diff
@@ -0,0 +1,183 @@
+ */
+
+#include <strings.h>
+#include "class/usbtmc/usbtmc_device.h"
```
`tusb.h` is the only file application should include from tinyusb.

hathach commented on src/class/usbtmc/usbtmc_device.h:
```diff
@@ -0,0 +1,133 @@
+extern usbtmc_response_capabilities_t const usbtmcd_app_capabilities;
+#endif
+
+bool usbtmcd_app_msgBulkOut_start(uint8_t rhport, usbtmc_msg_request_dev_dep_out const * msgHeader);
```
all device API must be prefixed with `tud_` and all lower case. Callback function (called from the stack) must has suffix `_cb`

hathach commented on src/device/usbd.c:
```diff
@@ -145,6 +145,22 @@ static usbd_class_driver_t const usbd_class_drivers[] =
+      .control_complete = usbtmcd_control_complete,
+      .xfer_cb          = usbtmcd_xfer_cb,
+      .sof              = NULL
+  },
```
I prefer to still keep it here.

hathach commented on src/device/usbd.c:
```diff
@@ -145,6 +145,22 @@ static usbd_class_driver_t const usbd_class_drivers[] =
+  // We maybe need to add subclass codes here, or a callback to ask if a driver can
+  // handle a particular interface.
+  {
+      .class_code       = TUSB_CLASS_APPLICATION_SPECIFIC,
```
we should keep the class code, since it is the only 0xFE for now, we will think how we should organize the subclass code later on.

hathach commented on src/class/usbtmc/usbtmc_device.h:
```diff
@@ -0,0 +1,133 @@
+void usbtmcd_init(void);
+
+/************************************************************
+ * USBTMC Descriptor Templates
```
all descriptors template must be prefixed with `TUD_` and follows the same naming convention as one in usbd.h
https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h#L164

I forgot the reason why I put them all in one place at usbd.h, look like they are better at their class driver. Anyways, please move it to usbd.h . I will check and possibly move them to their own class later on in a separate PR.

pigrew replied:
Ok, I'll move this right before finalizing the pull request (to minimize merge conflicts).

pigrew replied:
Ok, I'll move this right before finalizing the pull request (to minimize merge conflicts).

hathach commented on src/class/usbtmc/usbtmc.h:
```diff
@@ -0,0 +1,267 @@
+};
+
+typedef enum {
+  USBTMC_bREQUEST_INITIATE_ABORT_BULK_OUT      = 1u,
```
all UPPER case, you can remove the b

hathach commented on src/class/usbtmc/usbtmc.h:
```diff
@@ -0,0 +1,267 @@
+
+TU_VERIFY_STATIC(sizeof(usbtmc_read_stb_rsp_488_t) == 3u, "struct wrong length");
+
+typedef struct TU_ATTR_PACKET
```
`TU_ATTR_PACKET` seem to be a typo

pigrew commented:
> Thank you very much for yet another great PR. It is never be too soon to submit PR, you can constantly push the new code to update the PR. Even after merged, I did constantly changing the stack all around.
> 
> There is only API naming changes to keep thing consistency for now, though I will read some docs and try to get familiar with this class later on. Meanwhile, would you mind give me a bit of instruction on how to test this class: e.g application link, or simple instructions to use those app to test with.

Yes, I'l l do that. Which operating system do you use? The drivers are different for different OS. I can add a simple python script to talk to the device once the drivers are installed.

tannewt commented:
@pigrew Thank you for all of your contributions to TinyUSB. You mention 400b of type conversions. How are you measuring that? Is there a tool for it?

pigrew commented:
> @pigrew Thank you for all of your contributions to TinyUSB. You mention 400b of type conversions. How are you measuring that? Is there a tool for it?

I just look at the final binary size. :) At the end of a compile in STM32CubeIDE it runs objdump:
```
arm-none-eabi-gcc -o "test070_tusb_hid_msc.elf" @"objects.list"   -mcpu=cortex-m0 -T"C:\Users\nconrad\Documents\stm32_repos\tinyusb\hw\bsp\stm32f070rbnucleo\stm32F070rbtx_flash.ld" --specs=nosys.specs -Wl,-Map="test070_tusb_hid_msc.map" -Wl,--cref -Wl,--gc-sections -nostartfiles -static --specs=nano.specs -mfloat-abi=soft -mthumb -Wl,--start-group -lc -lm -Wl,--end-group
Finished building target: test070_tusb_hid_msc.elf
 
arm-none-eabi-objdump -h -S  test070_tusb_hid_msc.elf  > "test070_tusb_hid_msc.list"
arm-none-eabi-size   test070_tusb_hid_msc.elf 
   text	   data	    bss	    dec	    hex	filename
  35428	    676	   3580	  39684	   9b04	test070_tusb_hid_msc.elf
Finished building: default.size.stdout
 ```

Another nice thing is that the STM32CubeIDE GUI has a window where it shows the sizes of various objects:
![image](https://user-images.githubusercontent.com/14810130/65168273-282fd280-da12-11e9-9fd8-0c097e764d08.png)

pigrew commented:
The code is working much better now, but still has some bugs.

I added a python script to run which will communicate with the device, and check that it's functioning reasonably. It's pretty basic for the moment. It requires something like the "NI VISA" libraries to be installed, and python3.6+

I still have not cleaned up the function/variable names, but that's on the todo list once everything else settles.

hathach commented:
@pigrew could you please change the target branch from develop to master, it will make it easier to review.

pigrew commented:
> Thank you very much @pigrew for your contribuition. Though could you please change the target branch from develop to master, it got the EP patch work merged from develop already :)

Done!

But, keep in mind that I have a bunch of clean-ups left to do. Though, with the latest patch, it seems to be working decently.

hathach commented:
> > Thank you very much @pigrew for your contribuition. Though could you please change the target branch from develop to master, it got the EP patch work merged from develop already :)
> 
> Done!
> 
> But, keep in mind that I have a bunch of clean-ups left to do. Though, with the latest patch, it seems to be working decently.

Thanks, Let's me know when it is ready for merging, I will do a final review and testing.

pigrew commented:
> Thanks, Let's me know when it is ready for merging, I will do a final review and testing.
> 
> PS: I am a little bit busy with BLE work, may be slow in response to your work :)

Ah... BLE... I've done a bit with the TI CC2650... BLE is a nice protocol.

hathach commented:
> > Thanks, Let's me know when it is ready for merging, I will do a final review and testing.
> > PS: I am a little bit busy with BLE work, may be slow in response to your work :)
> 
> Ah... BLE... I've done a bit with the TI CC2650... BLE is a nice protocol.

I worked on Nordic nRF52 though, it is like wireless version of USB now, very handy to interface with phone. I mainly maintain this repo here https://github.com/adafruit/Adafruit_nRF52_Arduino if you are interested.

pigrew commented:
Ok, I've tested this driver combined with my other branches. The wakeup is working (passes the validator test).

pigrew commented:
With the latest changes, I think the class is ready for testing. The main functionality is working well. I'm worried a little bit about concurrency (I added locking for the state machine variable, but it isn't sufficient).

I added a python test script to the examples folder, which demonstrates the core functionality and some of the error handling.

I'm happy for it to be merged, or sit here in testing for a while. I'm sure that there will be issues that will come up.

pigrew commented:
I seem to have forgotten an important feature. The app layer has to be able to signal that it's busy, and have the class driver NAK the bulk-out endpoint. This will require another function for the app layer to call when it's ready. Delay state are not mentioned in the specification, so I think that a NAK is the proper response when the micro's internal buffers are full.

pigrew commented:
Having the app control the interface flow control is implemented. Also, I removed references to `rhport` from the application layer, as I think this only applies to host drivers? If anything, interface_id should be added, but for now I'm assuming a single interface.

hathach commented on src/class/usbtmc/usbtmc_device.h:
```diff
@@ -0,0 +1,123 @@
+// * (successful) tud_usbtmc_app_check_abort_bulk_in_cb
+// * (successful) usmtmcd_app_bulkOut_clearFeature_cb
+
+void tud_usbtmc_app_open_cb(uint8_t interface_id);
```
should we replace `tud_usbtmc_app` by simply `tud_usbtmc_` dropping **app* in API

pigrew replied:
I'm not sure. I've found it hard to keep track of which functions are for which direction of communication, and keeping app there makes it more clear that certain functions are for implementation by the app, not the class library.

I think other libraries fix this by using structs of function pointers to create a namespace, but the TUSB code at the moment doesn't do that.

I can go either way with this question.

hathach replied:
It is rather simple, if the API is invoked by the stack and application have to implement that function, it is callback and must end with `_cb()` for **callback**. Otherwise, it doesn't has `_cb()` e.g

tud_cdc_write() : implemented by stack
tud_cdc_rx_cb(): implemented by application since it is a callback from stack.

pigrew replied:
I'm going to change this to be as you suggested.

I usually read left-to-right, so looking for the _cb is less easy, as I need to read left to right to center.

However, there are more levels of complexity....
1. app calls class (tud_class_*)
2. class calls app (tud_class_*_cb)
3. stack calls class (tud_class_* or tud_class_*_cb); // Sometimes with _cb, other times without _cb?
3. class calls stack (usbd_*)

Number 1 and 2 and 3 have overlap. This somewhat impacts the app developer since auto-complete will suggest both.

hathach commented on src/class/usbtmc/usbtmc_device.h:
```diff
@@ -0,0 +1,123 @@
+bool tud_usbtmc_app_msgBulkOut_start_cb(usbtmc_msg_request_dev_dep_out const * msgHeader);
+// transfer_complete does not imply that a message is complete.
+bool tud_usbtmc_app_msg_data_cb( void *data, size_t len, bool transfer_complete);
+void usmtmcd_app_bulkOut_clearFeature_cb(void); // Notice to clear and abort the pending BULK out transfer
```
exposed API to application must be prefixed with `tud_` and all lower-case

hathach commented on examples/device/usbtmc/src/tusb_config.h:
```diff
@@ -0,0 +1,66 @@
+
+#define CFG_TUD_USBTMC           1
+#define CFG_TUD_USBTMC_ENABLE_INT_EP 1
+#define CFG_USBTMC_CFG_ENABLE_488    1
```
should be rename to `CFG_TUD_USBTMC_ENABLE_488` for consistency

pigrew commented on docs/concurrency.md:
```diff
@@ -0,0 +1,36 @@
+# Concurrency
```
How about if I let you handle moving things around?

I wonder how much is wrong, I wrote it before understanding the stack as well, so I don't know how correct it is.

hathach replied:
:ok_hand:

pigrew commented:
@hathach  Would you mind taking a look at the latest changes?

They are mostly cosmetic, but I did have some small updates to the testing script, and also changed the get_capabilities method to use a callback function instead of an extern struct.
